### PR TITLE
Output correct exit code when running tests

### PR DESCRIPTION
**Pull request by SUPERCILEX**
If the tests fail, we should exit with code 1. This lets you write a one line failure miner: `while ./run-tests.py --lab n; do :; done`.

### CODE REVIEW

**Comment by emichael** [comment]
Definitely agree that return code should be set, but this doesn't quite work.

The `main` method isn't always the last thing running in Java. In particular, the viz client is started as a daemon thread. The `System.exit` call at the end of `DSLabsTestCore` would immediately kill the viz client, meaning that the viz would never start.

Not sure what the right solution is. First thing that came to mind was a shutdown hook, but unfortunately the JVM runs them all concurrently and makes no guarantees about the order they're invoked. I use shutdown hooks to do other things, which a `System.exit` would similarly kill. I'll have to think about it.

**Comment by emichael** [comment]
Did a bit more digging, and I don't think setting the Java return code is feasible or the right way to do things. It's really only for JVM issues.

Best approach is probably to tee the output in python and watch for "FAIL" or maybe even "FAIL" followed by a newline and a row of = in the output. It's a bit brittle, but it's fine for now.

I don't have a ton of time to work on it right now. If you want to have a go, that would be great!

**Comment by SUPERCILEX** [comment]
Oops, forgot to tests that. :grimacing: I moved starting the viz debugger to a shutdown hook and removed a hook in the viz debugger itself that wouldn't run because we're already in a hook. I've tested `-z`, `--checks` and `--debug` which all seem to work.

File: framework/tst/dslabs/framework/testing/junit/BaseJUnitTest.java [hunk 1]
```diff
@@ -308,7 +308,7 @@ private void assertEndConditionValid() {
                 }
             }, "VizClient");
             thread.setDaemon(false);
-            thread.start();
+            Runtime.getRuntime().addShutdownHook(thread);
```
Comment (SUPERCILEX) [inline framework/tst/dslabs/framework/testing/junit/BaseJUnitTest.java:311]: Is there a case when more work will be done after we `throw new VizClientStarted()`? I'm wondering if this should instead be something like
```
vizThread.start();
Runtime.getRuntime().addShutdownHook(() -> vizThread.join());
```

**Comment by emichael** [comment]
That would work in this case, but it's way too hacky. Shutdown hooks aren't supposed to be long-lived. I'm not going to add any `System.exit` calls in Java. The only thing I would add is a Python returncode based on Java stdout/stderr.

File: framework/tst/dslabs/framework/testing/junit/BaseJUnitTest.java (hunk 1 above)
Comment (emichael) [reply framework/tst/dslabs/framework/testing/junit/BaseJUnitTest.java:311]: Moot not, but that wouldn't do anything. `vizThread` would get killed as soon as `System.exit` was called, and the shutdown hook would finish instantly.

**Comment by SUPERCILEX** [comment]
Just to make sure we're on the same page, calling exit does still run the shutdown hooks. If you'd rather not go that route, I'll look into doing it on the python side. :+1:

**Comment by emichael** [comment]
I understand. But running the viz client in a shutdown hook is a nonstarter. Take a look at the documentation for `addShutdownHook` and `exit`. Long-lived, complicated code shouldn't be run inside them.

**Comment by SUPERCILEX** [comment]
Alright, sounds good. I'll see about checking the output.

**Comment by SUPERCILEX** [comment]
Okey dokey, I've got some more attempts.

On the python side, I'm not a fan. :( You can't keep the stdout and stderr streams separate while preserving original output order, so that means you need to redirect stderr to stdout. It's not the end of the world, but still a bummer to be losing good redirection capabilities.

Back on the exit side, another solution that works is checking for the `VizClientStarted` exception and not exiting if it's present. This actually seems more correct than my first attempt because you don't really want a successful viz session to return an error code.

If neither of the approaches are good, I'm going to call it a bust.

**Comment by emichael** [comment]
Good point. You need to do it at a different level, by creating a file-like object that's a transparent pass-through to stdout/stderr and checks for "FAIL" along the way. The most bullet-proof way to do it is probably to implement `__getattr__` in such a way that it intercepts `write` and passes all other methods directly to the underlying object (stdout/stderr).

No need to work on this if you don't want to! I'll get around to it eventually.

**Comment by SUPERCILEX** [comment]
Yeah, sadly my python fu is not that good. :)

Would you mind sharing the issue with calling exit in java again? I'm not really seeing the downside if we ignore it when the visualizer is started.

**Comment by emichael** [comment]
It's just kind of hacky and brittle. Sure, it would work for now, but if later on I add some other daemon threads with some other functionality, it would break. Also, it's my impression that exit codes aren't really used in Java apps all that much. I'd rather keep the Java clean and keep the hacks on the python side.

**Comment by SUPERCILEX** [comment]
Thanks for elaborating. :)

**Pull request closed without merging**